### ISSUE TITLE
Move terrain generation to GPU

### ISSUE THREAD

**Issue by Inspirateur** [Inspirateur/riverbed#24]
For terrain generation to use 3D noise without being too slow (required for caves), we need it to be on GPU. 

Unfortunately it also means that we can't keep the nice abstraction provided by https://github.com/Inspirateur/noise-algebra, because that only works on CPU.

We could build back a nice abstraction on GPU but it has to be done incrementally because it's harder and I'm also less knowledgeable on shader code. It's a tough decision but it has to be done!

**Comment by Lemonzyy** [comment]
I don't know if you've considered these issues, but I think they're very important in terms of Riverbed's goals.

* Generating chunks on the GPU is great, but you want multiplayer. This means that if you're doing something like Minecraft where you can buy a server hosting solution, you'll need a server with a GPU which cost far more money. The same goes for people who host their own servers. Can they use an integrated GPU?
* If the server doesn't have a GPU, perhaps the clients can generate the map on their own? But then comes the problem of cheating and players who know the seed.

All this may or may not concern you, but I just wanted to share some of the issues I've been thinking about for my game.

**Comment by Inspirateur** [comment]
Thanks for your input :) These are interesting questions ! 
I do need to evaluate whether GPU generation is worth it or not, and maybe provide a fallback method of generation if no GPU is present 🤔 

Other solutions could include asking clients to generate the terrain indeed, and we could imagine cheating mitigation strategies such as validating random chunks ... 

Anyway I'll see that when I go over this issue :)

**Comment by nerdachse** [comment]
I just want to chime in and say: cheat prevention also depends on what kind of multiplayer you want to have.

Is it just friends / small scale, is it competitive?

If a few friends play together and one friend cheats: so what? Let the friends handle that.

If you give everyone a seed and let the clients generate the terrain and one client cheats they could potentially see far away valuable ores (just an example), but the server can still decide not to send them the info that an ancient dragon sleeps near those ores. Or the server can decide that it won't send any updates for that far away chunk... Movement and actions should still be server side validated, to disallow instant teleportation etc. pp.

Just my 0,05

**Comment by Inspirateur** [comment]
> I just want to chime in and say: cheat prevention also depends on what kind of multiplayer you want to have.

Yes, my stance on that is that being able to do multiplayer with friends is certainly valuable and is always the first step in development, but eventually I want to be able to have larger scale public server and that does require anti cheat measures, which we will figure out when we're there, meaning not now :p

**Comment by Inspirateur** [comment]
Added this internal crate https://github.com/Inspirateur/riverbed/tree/main/crates/riverbed_closest (what used to be "nd-intervals") for biome/plants selection. Hopefully this will allow us to blend custom noises per biomes together. It's just 1 small piece of the problem though.
